packages/vite/src/node/ssr/ssrExternal.ts

```diff
@@ -73,64 +97,46 @@ export function resolveSSRExternal(
       // which returns with '/', require.resolve returns with '\\'
       requireEntry = normalizePath(require.resolve(id, { paths: [root] }))
     } catch (e) {
+      try {
+        // no main entry, but deep imports may be allowed
+        const pkgPath = resolveFrom(`${id}/package.json`, root)
+        if (pkgPath.includes('node_modules')) {
+          ssrExternals.add(id)
+        } else {
+          depsToTrace.add(path.dirname(pkgPath))
+        }
+        continue
+      } catch {}
+
       // resolve failed, assume include
       debug(`Failed to resolve entries for package "${id}"\n`, e)
       continue
     }
+    // no esm entry but has require entry
     if (!entry) {
-      // no esm entry but has require entry (is this even possible?)
       ssrExternals.add(id)
-      continue
     }
-    if (!entry.includes('node_modules')) {
-      // entry is not a node dep, possibly linked - don't externalize
-      // instead, trace its dependencies.
-      depsToTrace.add(id)
-      continue
+    // trace the dependencies of linked packages
+    else if (!entry.includes('node_modules')) {
+      const pkgPath = resolveFrom(`${id}/package.json`, root)
+      depsToTrace.add(path.dirname(pkgPath))
     }
-    if (entry !== requireEntry) {
-      // has separate esm/require entry, assume require entry is cjs
+    // has separate esm/require entry, assume require entry is cjs
+    else if (entry !== requireEntry) {
       ssrExternals.add(id)
-    } else {
-      // node resolve and esm resolve resolves to the same file.
-      if (!/\.m?js$/.test(entry)) {
-        // entry is not js, cannot externalize
-        continue
-      }
-      // check if the entry is cjs
+    }
+    // externalize js entries with commonjs
```

Comment:
there changes are conflicting with the changes in https://github.com/vitejs/vite/pull/5197/files#diff-a6d6596967a09b96fa100c4b77480490b5ef95df88987ecbfa00ba1c868e4f53R97

rebased version is here: https://github.com/vitejs/vite/pull/5544

if you'd prefer to update this PR, I'm happy to close mine